src/charm.py

```diff
@@ -0,0 +1,553 @@
+            jobs=[
+                {
+                    "static_configs": [{"targets": [f"*:{PROMETHEUS_PORT}"]}],
+                }
+            ],
+        )
+        self._service_patcher = KubernetesServicePatch(
+            charm=self,
+            ports=[
+                ServicePort(name="pfcp", port=8805, protocol="UDP"),
+                ServicePort(name="bess-web", port=BESS_WEB_PORT),
+                ServicePort(name="prometheus-exporter", port=PROMETHEUS_PORT),
+            ],
+        )
+        network_attachment_definition_spec = {
+            "config": json.dumps(
+                {
+                    "cniVersion": "0.3.1",
+                    "type": "macvlan",
+                    "ipam": {"type": "static"},
+                    "capabilities": {"mac": True},
+                }
+            )
+        }
+        self._kubernetes_multus = KubernetesMultusCharmLib(
+            charm=self,
+            containers_requiring_net_admin_capability=[self._bessd_container_name],
+            network_attachment_definitions=[
+                NetworkAttachmentDefinition(
+                    metadata=ObjectMeta(name=ACCESS_NETWORK_ATTACHMENT_DEFINITION_NAME),
+                    spec=network_attachment_definition_spec,
+                ),
+                NetworkAttachmentDefinition(
+                    metadata=ObjectMeta(name=CORE_NETWORK_ATTACHMENT_DEFINITION_NAME),
+                    spec=network_attachment_definition_spec,
+                ),
+            ],
+            network_annotations=[
+                NetworkAnnotation(
+                    name=ACCESS_NETWORK_ATTACHMENT_DEFINITION_NAME,
+                    interface=ACCESS_INTERFACE_NAME,
+                    ips=[self._access_network_ip],
+                ),
+                NetworkAnnotation(
+                    name=CORE_NETWORK_ATTACHMENT_DEFINITION_NAME,
+                    interface=CORE_INTERFACE_NAME,
+                    ips=[self._core_network_ip],
+                ),
+            ],
+        )
+
+    def _write_bessd_config_file(
+        self,
+        upf_hostname: str,
+        upf_mode: str,
+        access_interface_name: str,
+        core_interface_name: str,
+        dnn: str,
+        pod_share_path: str,
+    ) -> None:
+        """Write the configuration file for the 5G UPF service."""
+        jinja2_environment = Environment(loader=FileSystemLoader("src/templates/"))
+        template = jinja2_environment.get_template(f"{CONFIG_FILE_NAME}.j2")
+        content = template.render(
+            upf_hostname=upf_hostname,
+            mode=upf_mode,
+            access_interface_name=access_interface_name,
+            core_interface_name=core_interface_name,
+            dnn=dnn,
+            pod_share_path=pod_share_path,
+        )
+        self._bessd_container.push(
+            path=f"{BESSD_CONTAINER_CONFIG_PATH}/{CONFIG_FILE_NAME}", source=content
+        )
+        logger.info(f"Pushed {CONFIG_FILE_NAME} config file")
```

Comment:
How so?

Reply:
f-strings are evaluated immediately by the interpreter, so if the log is disabled, it will still do potentially expensive computation. In this case the cost is negligible, but standardizing on one approach for logging prevent some errors like:

```
n = 99999999999
logger.debug(f"Factorial of {n} is {factorial(n)}")
```

Reply:
Hummmm we don't do any computation in logging here, it's always string manipulation. I'd optimise for readability instead of performance, especially in a charm.

Reply:
If that's the case should I revert the changes in my PR? I'd strive for having a consistent approach.

Reply:
I'm not sure I understand the consistency issue, if we use f-string everywhere we keep things consistent. 

Optimising for performance should be done when we have performance issues. Else, we should optimise for code simplicity and readability. Here we are taking a readability hit and optimising for a problem we don't have. If we truly want to improve performance in this charm code, we should start by checking where most of our computing/memory costs are. I expect those not to be in logging right now.